text/0002-new-architecture.md

```diff
@@ -0,0 +1,243 @@
+more Sentry deployments in different regions and for smaller clusters, some of these problems
+are becoming more pressing as they create complexity for each added installation.
+
+# Expected Workstreams
+
+This captures all expected streams of work that are in the scope of overhauling Sentry's
+architecture.
+
+## Processing Pipeline
+
+These are long running concerns with the processing pipeline that should be adressed.
+
+### Isolate Processing Pipeline out of Monolith
+
+The celery tasks, event manager and related functionality used by the processing pipeline is living
+alongside the rest of the Sentry code in the `sentry` monolith.  This means that it's quite chaotic
+and many unintended cross dependencies are regularly introduced into the code base.  As some of the
+processing pipeline wants to move out of the monolith for better scalability (in some extreme cases
+even move out of Python as host language) it would be helpful by moving all of the processing pipeline
+into a `sentry_pipeline` package.  It would be permissible for `sentry_pipeline` to import `sentry`
+but not the other way round.
+
+This will cause issues in the testsuite as the sentry test suite currently tries to create events
+regularly and pragmatic solutions for this need to be found.
+
+### Remove Pickle in Rabbit
+
+We still use pickle as serialization format for Celery which means that it's not possible for code
+outside of the Sentry monolith to dispatch tasks.  This is both an issue for the separation of
+`sentry` and `sentry_pipeline` (as import paths are changing) as well as it causes problems for
+dispatching and listening to tasks from Rust and other languages.
+
+### Remove HTTP Polling to Symbolicator
+
+The form of communication for event processing from the Python pipeline to Symbolicator (which is a
+Rust service) involves polling the symbolicator from Celery Python workers.  This has shown many issues
+in the past where the entire thing can tilt if the load on the symbolicators moves too far from the
+load assumptions of the Python polling workers.  The correct solution would be for tasks to be directly
+picked up by symbolicator.
+
+### Remove Celery and RabbitMQ for Kafka
+
+Our use of RabbitMQ is reaching the limits of what can be done with this system safely.  If we hit
+disk our entire pipeline crawls to a grind and we're no longer able to keep up with the traffic.
+As such we are already throttling how the events make it from kafka rabbit in extreme cases.  We are
+however relying on RabbitMQ to deal with the large variance of task execution times in the symbolication
+part of the pipeline.  For instance a JavaScript event can make it in the low milliseconds through
+the entire pipeline whereas a native event can spend up to 30 minutes or more in the pipeline in very
+extreme cases (completely cold caches against a slow symbol server).
+
+It should however still be possible to move this to a Kafka model where batches of tasks are redistributed
```

Comment:
From where I'm standing we have RabbitMQ for two reasons at the moment: one is that it does not have head of line blocking which makes it a pretty trivial choice for parts of the processing library where we are not quite sure how long something takes. The second reasons is that we build a lot of stuff on top of celery, particularly the cron based stuff.

The first reason is in fact a benefit but it can also be done with Kafka. The second reason is just a historic thing I believe. I'm not sure if we have actual cost or scaling benefits with RabbitMQ still but that's something I would like to explore.

Reply:
Particularly on the RabbitMQ side we currently operate with non durable, federated queues.

Reply:
One concern I have is that celery is a great abstraction for quick prototyping while Kafka is everything but. Moving every trivially low-scale task that doesn't have consistency requirements to Kafka today is much harder than creating a task that runs on worker-glob.

one thing we could do is to implement a celery broker on top of kafka. this would make it easier to gather operational experience with running certain product features on kafka vs rabbitmq, and could significantly sweeten the deal if we decided we're gonna ditch rabbitmq entirely for whatever reason.

Reply:
Have we considered a managed cloud service for tasks queues instead of RabbitMQ? [GCP Tasks](https://cloud.google.com/tasks) or [Pub/Sub](https://cloud.google.com/pubsub/docs/overview)? Here's a [comparison](https://cloud.google.com/tasks/docs/comp-pub-sub).

I can confirm as well that scaling of workers on RabbitMQ is simpler than on Kafka due to the 1 consumer / partition limit and scaling of consumer groups causing rebalancing.

Reply:
If we did want to do everything with Kafka [agree with @fpacifici that the parallel consumer should be investigated](https://github.com/getsentry/rfcs/pull/2#discussion_r936146948) for non-ordered workloads.

Reply:
I will say that I have no preference on if we want to use RabbitMQ (or a hosted queue) or Kafka. I can in general live with both but if we stay with RabbitMQ I would like to move us into a world where different codebases can use the queue and not just the monolith. Particularly also other languages.